sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala

```diff
@@ -46,6 +47,23 @@ import org.apache.spark.sql.connector.read.streaming.{Offset => OffsetV2}
 class OffsetSeqLog(sparkSession: SparkSession, path: String)
   extends HDFSMetadataLog[OffsetSeq](sparkSession, path) {
 
+  private val cachedMetadata = new ju.TreeMap[Long, OffsetSeq]()
+
+  override def add(batchId: Long, metadata: OffsetSeq): Boolean = {
+    val added = super.add(batchId, metadata)
+    if (added) {
+      // cache metadata as it will be read again
+      cachedMetadata.put(batchId, metadata)
+      // we don't access metadata for (batchId - 2) batches; evict them
```

Comment:
Note for reviewers: see MicroBatchExecution / ContinuousExecution around calling `source.commit()`. To provide the offset metadata to source side we read offset metadata for previous batch, which is probably written by this driver in previous batches.

Reply:
`TreeMap` seems overkill to me if we only cache at most 2 batches. How about using a 2-size array? We can put/get the elements by the index of (batchId % 2).

For example:

```scala
private val cachedMetadata = Array[OffsetSeq](null, null)

// put
cachedMetadata(batchId % 2) = metadata

// get
Option(cachedMetadata.get(batchId % 2)).orElse(super.get(batchId))
```

Reply:
TreeMap approach is pretty much intuitive compared to this, and the overhead should be pretty much trivial compared to FS ops. Let's not do over-engineering.

Reply:
Or I should just do this in MicroBatchExecution / ContinuousExecution instead - then the code doesn't need to be generalized. Let me have a time to revisit this.

Reply:
Hmm... it looks easier to change here and reflect to both micro-batch and continuous mode. Let's leave this as it is.

I would worry about not coupling metadata with batch ID - it could bring critical issue on committing to source when something is slightly messed up. e.g. In file stream source we delete source files based on the commit. We cannot trade-off between correctness and performance. I'd prefer safer approach unless it's identified to have a noticeable performance impact.

Reply:
> e.g. In file stream source we delete source files based on the commit. 

I don't understand why this could bring a correctness issue (I'm still very new to streaming so I may miss some use cases). The only problem I can imagine now is `get(batchId)` may lead to the correctness issue as `batchId` like 2 or 4 results in the same index. But this can also be addressed by saving the `batchId` at the same, e.g., `Array[(batchId, OffsetSeq)]`, which I think also addresses your concern for batchId coupling.

BTW, I did a minor performance test between these two ways, and seems it does have a performance impact:

```scala

// ~270ms
timeTakeMs {
  treeMap.put(1, "1")
  treeMap.put(2, "3")
  treeMap.put(3, "3")
  treeMap.headMap(2, true).clear()
}

// ~0ms
timeTakeMs {
  array(0) = "0"
  array(1) = "1"
  array(0) = "2"
}
```

Thoughts?

Reply:
```
import java.lang.{Long => JLong}
import java.util.{ArrayList, Collections, TreeMap}

def c(treeMap: TreeMap[Long, String]): Long = {
  val t1 = System.nanoTime()
  treeMap.put(1, "1")
  treeMap.put(2, "3")
  treeMap.put(3, "3")
  treeMap.headMap(2, true).clear()
  (System.nanoTime() - t1)
}

def d(treeMap: TreeMap[Long, String], idx: Long, value: String): Long = {
  val t1 = System.nanoTime()
  treeMap.put(idx, value)
  treeMap.headMap(idx - 2, true).clear()
  (System.nanoTime() - t1)
}

def experimentC(): Unit = {
  val latencies = new ArrayList[JLong]()
  val warmupCount = 1000000
  val runCount = 10000000

  (1 to warmupCount).foreach { _ =>
    val t = new java.util.TreeMap[Long, String]()
    c(t)
  }

  (1 to runCount).foreach { _ =>
    val t = new java.util.TreeMap[Long, String]()
    latencies.add(JLong.valueOf(c(t)))
  }

  java.util.Collections.sort(latencies)

  printLatencies(latencies)
}

def experimentD(): Unit = {
  val latencies = new ArrayList[JLong]()
  val warmupCount = 1000000
  val runCount = 10000000

  val t = new java.util.TreeMap[Long, String]()
  (1 to warmupCount).foreach { idx =>
    d(t, idx, idx.toString)
  }

  val t2 = new java.util.TreeMap[Long, String]()
  (1 to runCount).foreach { idx =>
    latencies.add(JLong.valueOf(d(t2, idx, idx.toString)))
  }

  printLatencies(latencies)
}

def printLatencies(latencies: ArrayList[JLong]): Unit = {
  val arraySize = latencies.size()
  val minIdx = 0
  val maxIdx = arraySize - 1
  val percentile50 = (arraySize * 0.5).toInt
  val percentile90 = (arraySize * 0.9).toInt
  val percentile99 = (arraySize * 0.99).toInt
  val percentile999 = (arraySize * 0.999).toInt
  val percentile9999 = (arraySize * 0.9999).toInt
  val percentile99999 = (arraySize * 0.99999).toInt
  val percentile999999 = (arraySize * 0.999999).toInt

  java.util.Collections.sort(latencies)

  Seq(minIdx, percentile50, percentile90, percentile99, percentile999, percentile9999, percentile99999, percentile999999, maxIdx).foreach { idx =>
    printLatency(latencies, idx)
  }  
}

def printLatency(latencies: ArrayList[JLong], idx: Int): Unit = {
  println(s"$idx th : ${latencies.get(idx) / 1000} microseconds = ${latencies.get(idx) / 1000000} milliseconds")
}

// experimentC()

/*
0 th : 0 microseconds = 0 milliseconds
5000000 th : 0 microseconds = 0 milliseconds
9000000 th : 0 microseconds = 0 milliseconds
9900000 th : 0 microseconds = 0 milliseconds
9990000 th : 1 microseconds = 0 milliseconds
9999000 th : 9 microseconds = 0 milliseconds
9999900 th : 37 microseconds = 0 milliseconds
9999990 th : 223 microseconds = 0 milliseconds
9999999 th : 53612 microseconds = 53 milliseconds
*/

experimentD()

/*
0 th : 0 microseconds = 0 milliseconds
5000000 th : 0 microseconds = 0 milliseconds
9000000 th : 0 microseconds = 0 milliseconds
9900000 th : 0 microseconds = 0 milliseconds
9990000 th : 0 microseconds = 0 milliseconds
9999000 th : 6 microseconds = 0 milliseconds
9999900 th : 25 microseconds = 0 milliseconds
9999990 th : 150 microseconds = 0 milliseconds
9999999 th : 57887 microseconds = 57 milliseconds
*/
```

2018 13-inch MBP, i7 quad-core 2.7Ghz

```
./bin/spark-shell --driver-memory 2g
...
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.0.1
      /_/

Using Scala version 2.12.10 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_191)
```

Still think this really matters?

Reply:
Even without warmup (commenting out),

```
// experimentC()

...
9999990 th : 1632 microseconds = 1 milliseconds
9999999 th : 60999 microseconds = 60 milliseconds

// experimentD()

...
9999990 th : 321 microseconds = 0 milliseconds
9999999 th : 35074 microseconds = 35 milliseconds
```

Reply:
Ok..I saw where the problem is with my test. You're right the latency is trivial.

I'm not against your solution here. But since we've reached here, I'd like to mention one more thing that TreeMap tends to produce the instant object `AscendingSubMap` for each batch while Array doesn't. Although, It might also be trivial.

Reply:
This is another sort of micro-optimization; realistic latency of micro-batch is 1s+ (doesn't matter if we consider very tight micro-batch, like 500ms) and we worry about creating "an" object per such period which will be marked as "unused" after couple of batches.

This is the clear example why micro-optimization is bad without understanding full context - optimization should evaluate about the impact and proceed only when it contributes at least 1% (I'd rather not even concern about 1% though if the code is more intuitive).

Reply:
Thanks for bringing the details, it makes sense to me. But please also note that I wouldn't raise such optimization changes if the `TreeMap` is already an existing implementation. However, for a PR, I think it's good to have more inputs regardless of the final decision.

Reply:
The reason I am concerning on the input on possible micro-optimization is that it changes the viewpoint of the codebase to performance and doesn't evaluate based on other viewpoints. There're bunch of points to discuss on this change, is this better to be put to OffsetSeq or better to place it on both MicroBatchExecution/ContinuousExecution (and why), which code would be most intuitive to maintainers or contributors, etc. Once the performance perspective comes in and gains traction, other perspectives tend to be disregarded. That should be just one of viewpoints, and shouldn't be overvalued.

Reply:
I know your concern. I like those general optimization ideas too instead of hacks(micro-optimization). Here the argument may be that you think the `Array` way is not intuitive enough. Well, I'm actually neutral on this as it's also not too complex as it is. Let's move on since I already +1 for the current solution :)

Reply:
I've gone through this thread and changing the current solution would introduce quite some non-trivial code structures and the gain would be not significant from memory point of view so +1 to keep it as-is.